Source/Core/Core/PowerPC/Jit64/Jit.cpp

```diff
@@ -690,6 +690,38 @@ const u8* Jit64::DoJit(u32 em_address, PPCAnalyst::CodeBuffer* code_buf, JitBloc
     }
   }
 
+  // If the block depends on an input register which looks like a gather pipe or MMIO related
+  // constant, guess that it is actually a constant input, and specialize the block based on this
+  // assumption. This happens when there are branches in code writing to the gather pipe, but only
+  // the first block loads the constant.
+  // Insert a check at the start of the block to verify that the value is actually constant.
+  // This can save a lot of backpatching and optimize gather pipe writes in more places.
+  if (js.noSpeculativeConstantsAddresses.find(js.blockStart) ==
+      js.noSpeculativeConstantsAddresses.end())
+  {
+    for (auto i : code_block.m_gpr_inputs)
+    {
+      u32 compileTimeValue = PowerPC::ppcState.gpr[i];
+      if (PowerPC::IsOptimizableGatherPipeWrite(compileTimeValue) ||
+          PowerPC::IsOptimizableGatherPipeWrite(compileTimeValue - 0x8000) ||
+          compileTimeValue == 0xCC000000)
+      {
+        CMP(32, PPCSTATE(gpr[i]), Imm32(compileTimeValue));
+        FixupBranch failure = J_CC(CC_NZ, true);
+        SwitchToFarCode();
```

Comment:
you only need to generate the far code for the first value which matches. Any other failures can reuse the same farcode.